Composer/packages/server/src/models/storage/localDiskStorage.ts

```diff
@@ -74,4 +76,35 @@ export class LocalDiskStorage implements IFileStorage {
   async rename(oldPath: string, newPath: string): Promise<void> {
     return await rename(oldPath, newPath);
   }
+
+  async zip(source: string, out: string, res) {
+    const archive = archiver('zip');
+
+    archive.on('error', function(err) {
+      res.status(500).send({ error: err.message });
+    });
+
+    res.attachment('tmp-archive.zip');
+
+    archive.pipe(res);
+
+    // We're selectively adding specific directories/files to the archive.
+    // If a user has ejected the runtime into the path, we don't want to include
+    // these files into the archive
+    [
+      p.format({ dir: `${source}/dialogs/` }),
+      p.format({ dir: `${source}/language-understanding/` }),
+      p.format({ dir: `${source}/language-generation/` }),
+      p.format({ dir: `${source}/settings/` }),
+    ].forEach(directory => {
+      archive.directory(directory, directory.split(source)[1]);
+    });
+
+    const files = await glob('*.dialog', { cwd: source, dot: true });
+    files.forEach(file => {
+      archive.file(p.format({ dir: `${source}/`, base: `${file}` }), { name: p.basename(file) });
+    });
+
+    archive.finalize();
+  }
```

Comment:
A way to decouple this from express is to accept a callback function.

```ts
async zip(source: string, out: string, cb: (archiver: Archiver) => void {
  const archive = archiver('zip');

  cb(archiver)
}

// caller with a writeable stream in scope
zip('some-directory', (archiver) => {
  archiver.on('error', (err) => {
    res.status(500).send({ error: err.message });
  });

  archiver.pipe(res);
});
```

This decouples the archiving from what to do with the archive and would allow us to have an express archiver as well as a fs archiver.

Reply:
donezo